File: kimi-k3/reference/python/kimi_k3_ref/short_conv.py
"""Causal short convolution + SiLU used before KDA q/k/v projections.

Educational implementation of FLA ``ShortConvolution`` for a single sequence
(kernel_size=4 in Kimi K3). Depthwise: each channel has its own kernel.
"""

from __future__ import annotations

import numpy as np


def silu(x: np.ndarray) -> np.ndarray:
    x = np.asarray(x, dtype=np.float32)
    return (x / (1.0 + np.exp(-x))).astype(np.float32)


def short_conv1d_silu(
    x: np.ndarray,
    weight: np.ndarray,
    bias: np.ndarray | None = None,
) -> np.ndarray:
    """
    Parameters
    ----------
    x : (B, T, C) float32
    weight : (C, K) depthwise conv kernels (channel, kernel) — time-reversed
             so weight[..., 0] multiplies the oldest padded sample.
             Equivalent to nn.Conv1d groups=C with weight shape (C,1,K)
             flattened to (C,K).
    bias : optional (C,)

    Causal: left-pad with K-1 zeros, then for each t:
        y[t,c] = sum_{i=0}^{K-1} weight[c,i] * x_pad[t+i, c]
    Then SiLU.
    """
    x = np.asarray(x, dtype=np.float32)
    weight = np.asarray(weight, dtype=np.float32)
    bsz, t, c = x.shape
    k = weight.shape[1]
    assert weight.shape[0] == c
    # pad left
    x_pad = np.pad(x, ((0, 0), (k - 1, 0), (0, 0)), mode="constant")
    y = np.zeros_like(x)
    for i in range(k):
        y += x_pad[:, i : i + t, :] * weight[None, None, :, i]
    if bias is not None:
        y = y + np.asarray(bias, dtype=np.float32)[None, None, :]
    return silu(y)

File: kimi-k3/rust/src/short_conv.rs
#[inline]
fn silu(x: f32) -> f32 {
    x / (1.0 + (-x).exp())
}

/// Causal depthwise short conv + SiLU. `x`: (b,t,c), `weight`: (c,k).
pub fn short_conv1d_silu(
    x: &[f32],
    batch: usize,
    time: usize,
    channels: usize,
    weight: &[f32],
    kernel: usize,
    out: &mut [f32],
) {
    assert_eq!(x.len(), batch * time * channels);
    assert_eq!(weight.len(), channels * kernel);
    assert_eq!(out.len(), x.len());
    out.fill(0.0);
    for b in 0..batch {
        for t in 0..time {
            for c in 0..channels {
                let mut acc = 0.0f32;
                for i in 0..kernel {
                    let src_t = t as isize - (kernel as isize - 1 - i as isize);
                    let xv = if src_t >= 0 {
                        x[((b * time + src_t as usize) * channels) + c]
                    } else {
                        0.0
                    };
                    // weight[..., 0] multiplies oldest padded sample
                    // For padded index p = t + i in pad scheme (pad left k-1):
                    // y += x_pad[:, i:i+t] * weight[:,:,i]
                    // src at time t uses pad position t+(k-1) for i=k-1 (current)
                    // Our loop: i=0 -> oldest = t-(k-1)
                    acc += xv * weight[c * kernel + i];
                }
                out[(b * time + t) * channels + c] = silu(acc);
            }
        }
    }
}
